pkg/reconciler/taskrun/metrics.go

```diff
@@ -0,0 +1,271 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	trDuration = stats.Float64(
+		"taskrun_duration_seconds",
+		"The taskrun's execution time in seconds",
+		stats.UnitDimensionless)
+	trDistribution = view.Distribution(10, 30, 60, 300, 900, 1800, 3600, 5400, 10800, 21600, 43200, 86400)
+
+	prTRDuration = stats.Float64(
+		"pipelinerun_taskrun_duration_seconds",
+		"The pipelinerun's taskrun execution time in seconds",
+		stats.UnitDimensionless)
+	prTRLatencyDistribution = view.Distribution(10, 30, 60, 300, 900, 1800, 3600, 5400, 10800, 21600, 43200, 86400)
+
+	trCount = stats.Float64("taskrun_count",
+		"number of taskruns",
+		stats.UnitDimensionless)
+
+	runningTRsCount = stats.Float64("running_taskruns_count",
+		"Number of taskruns executing currently",
+		stats.UnitDimensionless)
+
+	podLatency = stats.Float64("taskruns_pod_latency",
+		"scheduling latency for the taskruns pods",
+		stats.UnitMilliseconds)
+)
+
+type Recorder struct {
+	initialized bool
+
+	task        tag.Key
+	taskRun     tag.Key
+	namespace   tag.Key
+	status      tag.Key
+	pipeline    tag.Key
+	pipelineRun tag.Key
+	pod         tag.Key
+}
+
+func NewRecorder() (*Recorder, error) {
+	r := &Recorder{
+		initialized: true,
+	}
+
+	task, err := tag.NewKey("task")
+	if err != nil {
+		return nil, err
+	}
+	r.task = task
+
+	taskRun, err := tag.NewKey("taskrun")
+	if err != nil {
+		return nil, err
+	}
+	r.taskRun = taskRun
+
+	namespace, err := tag.NewKey("namespace")
+	if err != nil {
+		return nil, err
+	}
+	r.namespace = namespace
+
+	status, err := tag.NewKey("status")
+	if err != nil {
+		return nil, err
+	}
+	r.status = status
+
+	pipeline, err := tag.NewKey("pipeline")
+	if err != nil {
+		return nil, err
+	}
+	r.pipeline = pipeline
+
+	pipelineRun, err := tag.NewKey("pipelinerun")
+	if err != nil {
+		return nil, err
+	}
+	r.pipelineRun = pipelineRun
+
+	pod, err := tag.NewKey("pod")
+	if err != nil {
+		return nil, err
+	}
+	r.pod = pod
+
+	err = view.Register(
+		&view.View{
+			Description: trDuration.Description(),
+			Measure:     trDuration,
+			Aggregation: trDistribution,
+			TagKeys:     []tag.Key{r.task, r.taskRun, r.namespace, r.status},
+		},
+		&view.View{
+			Description: prTRDuration.Description(),
+			Measure:     prTRDuration,
+			Aggregation: prTRLatencyDistribution,
+			TagKeys:     []tag.Key{r.task, r.taskRun, r.namespace, r.status, r.pipeline, r.pipelineRun},
+		},
+		&view.View{
+			Description: trCount.Description(),
+			Measure:     trCount,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{r.status},
+		},
+		&view.View{
+			Description: runningTRsCount.Description(),
+			Measure:     runningTRsCount,
+			Aggregation: view.LastValue(),
+		},
+		&view.View{
+			Description: podLatency.Description(),
+			Measure:     podLatency,
+			Aggregation: view.LastValue(),
+			TagKeys:     []tag.Key{r.task, r.taskRun, r.namespace, r.pod},
+		},
+	)
+
+	if err != nil {
+		r.initialized = false
+		return r, err
```

Comment:
We wanted to return the `nil`, however everytime in before calling the metrics methods in controller flow we need to check for `nil`, right? instead of doing that way we thought to add that check while calling metrics methods. Now trade here is the method has to check if the instance is actually `initialized` which is either not that great but it avoid many if check scatter accross caller places.